keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+  ...
+```
+
+Note that, in this design, groups of validations that should be enforced
+differently must be split out across multiple policy definitions.
+
+xref:
+
+- https://open-policy-agent.github.io/gatekeeper/website/docs/next/violations/
+
+#### Informational type checking
+
+This is complicated by:
+
+- Version skew
+- CRDs
+- Aggregated API servers
+
+Problem examples:
+
+| Problem                                                | Summary                                     |
+| ------------------------------------------------------ | ------------------------------------------- |
+| version skew: ephemeralContainers case                 | New pod field, need to be able to validate in same was containers and initContainers if field exists and is populated |
+| version skew: Migration from annotation to field       | Need to be able to validate annotation (if present) or field (if it exists and is populated) |
+| CRD is deleted                                         | Nothing to type check against, but also means there are no coresponding custom resources |
+| CRD is in multiple clusters, but schema differs        | If policy author is aware of the schema variations, can they write policies that work for all the variations? |
+| Validation of an aggregated API server type            | Main API server does not have type definitions |
+
+Due to these complications, we have decided to evalute CEL expressions
+dynamically. Informational type checking will be provided (except for aggregated
+API server types), but will be surfaced only as warnings. See "Alternatives
+Considered" section for details of all the alternatives we reviewed when
+selecting this approach.
+
+Type checking is still performed for all expressions where a GVK can be matched
+to type check against, resulting in warnings, e.g.:
+
+```yaml
+...
+status:
+  expressionWarnings:
+    - expression: "object.foo"
+      warning: "no such field 'foo'"
+```
+
+#### Failure Policy
+
+For in-process validation there is no remote request, so errors should be
+deterministic. We believe this significantly reduces the risk of "fail closed"
+admission control as compared to webhooks.
+
+Because failure policy is most often selected based on the need to guarantee
+enforcement, we will default failure policy to "fail" and allow it to
+be configured on a per-policy basis:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+spec:
+  ...
+  failurePolicy: Ignore # The default is "Fail"
+  validations:
+    - expression: "object.spec.xyz == params.x"  
+```
+
+#### Safety measures
+
+To prevent clusters from being put into a unustable state that cannot be
+recoverd from via the API, admission webhooks are not allowed to match
+`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.
+
+We will extend this approach:
+
+- `ValidatingAdmissionPolicy` cannot match `ValidatingAdmissionPolicy`
```

Comment:
I think you should also outlaw matching param objects or bindings!

Reply:
Good catch! Fixed.